In [224]:
import pandas as pd
import numpy as np

In [225]:
#Loading the dataset as a Pandas DataFrame
df = pd.read_csv('/kaggle/input/creditcardfraud/creditcard.csv')

# Exploratory Data Analysis (EDA)

In [226]:
# To see first 5 datapoints(rows)
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

* Here, we can see all columns are in numeric (float). Actually these columnas are the information of each transaction details. And these columns are converted to numeric (float) by using PCA (Principal Component Analysis) method, for not to share transaction details directly.
* Principal Component Analysis (PCA): is a statistical procedure  that uses an orthogonal transformation that converts a set of correlated variables to a set of uncorralated variables (PCA is the most widely used tool in exploratory  data analysis and in machine learning for predective models).

In [227]:
# To see last 5 datapoints (rows)
df.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [228]:
# To check no.of rows and columns
df.shape

(284807, 31)

In [229]:
# To see dataset Information 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [230]:
# To check missing values in each column
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

Here, we can see that there are no missing values in this dataset.

In [231]:
# To check the distribution of legit (legitimate) transactions and fradulent transactions
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

We can see that this dataset is highly unbalenced.

Here,
* 0 - Legitimate transactons
* 1 - Fradulent transactions

In [232]:
# Separating the data for analysis 
legit = df[df.Class == 0]
fraud = df[df.Class == 1]

In [233]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [234]:
# To check statistical measures (information) of the data 
# Amount column
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [235]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [236]:
# compare the values for both transcations 
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

 Now I will do 'Under-Sampling' to balance the unbalanced data
 
 # Under-Sampling

Build a sample dataset containing similar distribution of normal transactions and Fradulent transactions.

Number of Fradulent transcations is 492

In [237]:
legit_sample = legit.sample(n=492,random_state=2)
legit_sample.shape

(492, 31)

In [238]:
# now concatenating two DataFrames 
new_dataset = pd.concat([legit_sample,fraud],axis=0) # concat rows wise
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
188671  128077.0 -3.288944 -4.809848 -1.231524  6.041078 -1.948400  1.772856   
190546  128878.0  2.020493 -1.000658 -1.039495 -0.453275 -0.737034 -0.433760   
46318    42728.0  0.883694 -0.761362  0.928801  1.389779 -0.730351  1.228938   
267636  162855.0 -0.072377  0.735400 -2.211240 -2.153156  3.556343  2.781633   
189610  128481.0 -0.264285  0.990040 -0.643148 -0.984799  0.813840  0.033159   

              V7        V8        V9  ...       V21       V22       V23  \
188671  2.177367  0.561664 -1.859951  ...  1.352295  0.869342  2.760665   
190546 -0.587955 -0.148476 -0.491796  ...  0.024865  0.619987 -0.041074   
46318  -0.771123  0.432439  1.196648  ... -0.036445 -0.104183 -0.321204   
267636  1.142571  0.602748 -0.561615  ...  0.256899  0.796234 -0.159167   
189610  0.536661  0.483326 -0.368426  ... -0.246360 -0.708310 -0.024761   

             V24       V25       V26       V27       V28   Amount  Class  
188671 -0.514612  0.114622  0.531044 -0.533325 -0.548096  1206.14      0  
190546 -0.264674 -0.025601  0.910031 -0.051887 -0.063195    67.95      0  
46318  -0.812608  0.547630 -0.273658  0.055533  0.039590   150.00      0  
267636  0.752505 -0.263957  0.108374  0.406915  0.263924    24.00      0  
189610 -1.457310 -0.325381  0.213825  0.115022  0.010990    14.28      0  

[5 rows x 31 columns]

In [239]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [240]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [241]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      95169.597561 -0.019460  0.014204 -0.073895  0.051097  0.102674   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.022626  0.014979 -0.012521 -0.014322  ... -0.079576  0.019515   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.038665 -0.004756  0.042502  0.009427  0.014371 -0.003561 -0.021240   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       75.156280  
1      122.211321  

[2 rows x 30 columns]

here, we can see that the nature of the new_dataset and original dataset are almost same (familiar) and got good samples.

# Spliting the data into Features and target

In [242]:
x = new_dataset.drop(['Class'],axis=1)
y = new_dataset['Class']

In [243]:
print(x.shape)
print(y.shape)

(984, 30)
(984,)


# Split the data into training and testing data

In [244]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2, stratify=y, random_state=2)

In [245]:
print(x.shape,x_train.shape,x_test.shape)

(984, 30) (787, 30) (197, 30)


# Model training

In [246]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

In [247]:
# training the LogisticRegression model with Training data
model.fit(x_train,y_train)

LogisticRegression()

Model Evaluation

In [248]:
# To check the accuracy score on training data
y_train_pred = model.predict(x_train)

In [249]:
from sklearn.metrics import accuracy_score

print('model accuracy score on training data : {:.4f}'.format(accuracy_score(y_train_pred,y_train)))

model accuracy score on training data : 0.9149


In [250]:
# Check the accuracy score on test data
y_pred = model.predict(x_test)

print('model accuracy score on test data : {:.4f}'.format(accuracy_score(y_pred,y_test)))

model accuracy score on test data : 0.9137


* The accuracy score on training data (0.9149) and accuracy score on test data (0.9137) are same. So, there is no underfitting and overfitting problem in this is model.
* The prediction of this model is very accurate on both training and test data. 
* so, this is the best model for predictions of 'credit card fraud detection'.